qt: Add Window menu

promag opened the pull request:
Overall this PR does the following:
 - add top level menu Window
 - add Minimize and Zoom actions to Window menu
 - move Sending/Receiving address to Window
 - remove Help->Debug window
 - add one menu entry for each debug window tab

This removes the access to address book from the File menu.

With wallet support:
<img width="522" alt="screenshot 2018-12-11 at 00 33 05" src="https://user-images.githubusercontent.com/3534524/49770451-5bec0800-fcdc-11e8-91d6-f8f850ead92d.png">

Without wallet support:
<img width="593" alt="screenshot 2018-12-11 at 12 55 21" src="https://user-images.githubusercontent.com/3534524/49802183-19f6ac80-fd44-11e8-9973-36fcfb4f129e.png">

promag committed:
qt: Add Window menu

ken2812221 reviewed (commented):
Concept ACK

hebasto commented:
Concept ACK.
Could move "Backup Wallet..." from File menu to Wallet menu?

From [Apple's Human Interface Guidelines](https://developer.apple.com/design/human-interface-guidelines/macos/menus/menu-bar-menus/):

> Provide a Window menu even if your app has only one window. Include the Minimize and Zoom menu items so people using Full Keyboard Access can invoke these functions using their keyboard.

promag commented:
Removed the changes to the "Settings" menu.

laanwj commented:
I don't know what the conventions around this are, nowadays, but Concept ACK.

MeshCollider commented:
Concept ACK

promag commented:
Thanks @jonasschnelli, I'll see if others also toggle.

I have to test in windows.

promag commented:
@jonasschnelli on macos, if the active window is minimized then "Minimize" action is disabled. And "Zoom" action toggles between maximized and normal.

Also, some software add the main window to this menu, and I think we could too.

Sjors commented:
@hebasto wrote:

> Could move "Backup Wallet..." from File menu to Wallet menu?

Why? Loading, saving and backing up the wallet seem like document / File operations.

Moving message signing and verification from File to Window would make more sense to me.

Even more clear IMO, but haven't thought very deeply about it yet, is to have a _Wallet_ top level menu. In that case I'd be more in favour of adding Backup into that, as well as sign/verify, Sending/Receiving address. The Window menu would be used for Minimize, Debug Window.

I would also be fine with dropping the "Address Book" level and just having direct menu items for Sending and Receiving addresses.


Definitely out of scope, but I would still like to see each wallet represented as a window (rather than the little dropdown we have now). However I wouldn't want closing a window to cause unloading a wallet, so it's not obvious how this would work.

hebasto commented:
IMO, it would be better to avoid using native widgets ([Native Widgets vs Alien Widgets](https://doc.qt.io/qt-5.9/qwidget.html#native-widgets-vs-alien-widgets)) and related methods (`QWidget::windowHandle()`) and objects (`QWindow`).

@jonasschnelli 
> I think the "minimize" entry should toggle to "maximise" once minimized or at least maximises when selecting "minimize" in already minimized state.

@promag 
> on macos, if the active window is minimized then "Minimize" action is disabled. And "Zoom" action toggles between maximized and normal.

On macOS, if the active window is minimized the both "Minimize" and "Zoom" actions should be disabled. That is not the case for 3f779b3b99b674b7beecc9a30fa37bda1d722387.

hebasto commented on src/qt/bitcoingui.cpp:
```diff
@@ -396,11 +394,32 @@ void BitcoinGUI::createMenuBar()
+    });
+#ifdef Q_OS_MAC
+    window_menu->addAction(tr("Zoom"), [this] {
+        if (windowHandle()->windowState() == Qt::WindowMaximized) {
```
The same:
`if (isMaximized()) {`

promag replied:
Same as above.

promag commented on src/qt/bitcoingui.cpp:
```diff
@@ -396,11 +394,32 @@ void BitcoinGUI::createMenuBar()
+    QMenu* window_menu = appMenuBar->addMenu(tr("&Window"));
+    QAction* minimize_action = window_menu->addAction(tr("Minimize"), this, &QMainWindow::showMinimized, QKeySequence(Qt::CTRL + Qt::Key_M));
+    QObject::connect(windowHandle(), &QWindow::windowStateChanged, [this, minimize_action] {
+        minimize_action->setEnabled(windowHandle()->windowState() != Qt::WindowMinimized);
```
Does not apply now since it's using `QWindow` now.

promag commented:
Updated after feedback, thank you! Also updated screenshot. Windows/linux feedback would be awesome.

promag committed:
qt: Remove ellipsis from sending/receiving addresses

Considering https://stackoverflow.com/a/637708 the ellipsis in these
menu actions should be removed.

promag commented:
Pushed a commit to remove ellipsis from the sending/receiving addresses menu actions. Please see rationale in https://stackoverflow.com/a/637708.

Sjors commented:
tACK 6c72f91
A few nits, but don't let that stop anyone from merging:
*  you could amove Sign & Verify Message to Window (they do need ellipses)
* I don't find the Zoom item very useful, but don't mind it either
* when  `./configure --disable-wallet` or `disablewallet=1` you could still show Information, Console, Network Traffic and Peers in the Window menu. There may be some accessibility advantage to that.

promag commented:
> * you could move Sign & Verify Message to Window (they do need ellipses)

I'll leave that for later after we decide about top level menu "Wallet".

> * I don't find the Zoom item very useful, but don't mind it either (I'm just not a Bitcoin Core Window Maximalist)

Minimize and Zoom operate on the focus window, which can be the main window, address book dialogs, debug window..

> * when  `./configure --disable-wallet` or `disablewallet=1` you could still show Information, Console, Network Traffic and Peers in the Window menu.

I'll update with this suggestion, thanks @Sjors.

promag committed:
qt: Allow to inspect RPCConsole tabs

Sjors commented:
tACK 9ea38d0

```
git range-diff `git merge-base --all HEAD 6c72f91`...6c72f91 HEAD~3...HEAD
```

<img width="629" alt="schermafbeelding 2018-12-11 om 17 39 04" src="https://user-images.githubusercontent.com/10217/49816643-9dc58e80-fd6e-11e8-919d-fa451cf1ebe0.png">

promag commented:
@ken2812221 do you think you have the time to test this?

MeshCollider commented:
Tested ACK https://github.com/bitcoin/bitcoin/pull/14573/commits/9ea38d022281713e7f79a219b37651ac5648d695 on windows 10

I'm a little confused about the "Main Window" option, isn't the menu part of the main window implying that it is already focused? Also perhaps "Restore" should toggle between "Maximise" & "Restore"?

![image](https://user-images.githubusercontent.com/3211283/49977319-c8039000-ffaa-11e8-98fd-440a6c3db6f6.png)

promag commented:
@MeshCollider thanks! 

This was inspired on macos since it has a menu bar per application, not per window. Forgot that in windows it has menu bar in each window..

So next changes will be:
 - first section are actions on the current window (focused window on macos)
 - then comes the main windows — we only allow 1 main window for now — and this is used to bring it to front (I'll remove in windows, maybe linux too?)
 - then comes some dialogs — address book for now
 - then comes the debug dialogs

Sjors commented:
Rather than explicitly handling this per OS, does QT have any way to abstract that?

promag commented:
Not that I'm aware of.

MarcoFalke commented:
This does not compile because the addAction function was only introduced in Qt 5.6.

Empact commented:
An example of the failure: https://travis-ci.org/bitcoin/bitcoin/jobs/468706526

molxyz commented:
Hello, I also have a question on that button `Main Window` on windows bitcoin. If it has no purpose please remove that button before the official release to avoid confusions for windows users.

Thank you.

Please remove this button: 

![image](https://user-images.githubusercontent.com/21127776/56612154-99916380-65e1-11e9-8bac-3664b8a308b3.png)

promag commented:
@molxyz from the above https://github.com/bitcoin/bitcoin/pull/14573#issuecomment-447184073

> * then comes the main windows — we only allow 1 main window for now — and this is used to bring it to front (I'll remove in windows, maybe linux too?)

Sorry for forgetting about this, but at this stage I'm afraid it won't be removed for 0.18.

molxyz commented:
@promag Hi, I'll be happy to run tests for RC5.  Thanks.
